### ISSUE TITLE
Documentation feedback

### ISSUE THREAD

**Issue by Timmmm** [rrousselGit/riverpod#326]
Hi, I just started learning Riverpod, thought you might appreciate some feedback on the docs. In no particular order:

1. It's a really neat idea. I like it.
2. None of the links in the docs work, e.g. they point here which is 404: https://pub.dev/documentation/riverpod/latest/all/Provider-class.html
3. There are loads of different providers and there's nowhere in the docs where they are all listed and the differences between them are clarified. For example [here](https://riverpod.dev/docs/concepts/providers) it says:

> Provider is the most basic of all providers. It exposes an object that never changes.
We could replace Provider with other providers like StreamProvider or StateNotifierProvider, to change how the value is interacted with.

But I don't think there's anywhere in the docs where it really lists them in a nice table like this:

| Provider Type | Description |
|----------------|--------------|
| `Provider` | Basic provider for values that never change. If you try to `watch()` it then ??? happens. |
| `FutureProvider` | Provider for a `Future` where only the completion status of the future changes. This is useful for async tasks that happen exactly once, for example loading config files. |
| `StreamProvider` | Provider where the state is a stream. `watch()`ing the stream causes the widget to be rebuilt when a new value appears on the stream. |
| `StateProvider` | Not exactly sure about this? Does `watch()` get notified when the state changes? If so how is it different to `StateNotifierProvider`?
| `StateNotifierProvider` | ???|

etc.

4. On the [`StateNotifierProvider`](https://pub.dev/documentation/riverpod/latest/riverpod/StateNotifierProvider-class.html) it says:

> This provider is used in combination with `package:state_notifier`.

But none of the examples show that and all of the classes seem to be in `riverpod_flutter`?

5. There aren't any examples for providers with mutable state, where I would manually trigger an update. For example:

```
class AppState {
  int x;
  int y;
  int z;

  void doThing() {
    this.x += 1;
    notifyWatchersSomehow();
  }
}
```

Maybe there's something similar to Flutter's `setState()`? It seems like an odd omission - I'd expect 90% of app state to be big heterogenous collections of mutable state, rather than simple `int counter` things that are always shown as examples. The `TodosNotifier` example is better but it still completely reassigns the entire state object to trigger updates, which I doubt is a typical use.

[This example](https://github.com/lohanidamodar/flutter_worldtime/blob/master/lib/presentation/notifiers/clocks_notifier.dart) seems to essentially trigger an update via `state = state` but it's not an official example so I'm not sure if that's the best option.

6. Sometimes you do `watch(provider)` and sometimes `watch(provider.state)`. Maybe I missed it but I didn't see any explanation of why.
7. I'd appreciate some more explanation of how it all works up front. I suspect it's pretty much impossible to use this without understanding at least vaguely how it works, and the current guide is a load of "if you want this, then do this" rather than "when you call `watch()` it does this".

Hope that helps! Like I said, it seems like a great project and I really want to learn how to use it but there's definitely a slightly inconvenient focus on "how to do a counter" and "how to do a todo list" rather than "this is how it works" and "how to do an app with state that is more complicated than just a list or an integer".

Thanks!

**Comment by rrousselGit** [comment]
Thanks a lot for the feedback! I really appreciate 😄 

Overall I agree with what you said. The main issue is finding time to edit the documentation at the moment. I'm quite busy with writing a devtool.

Good call on the broken pub links. I'll fix those





> 1. On the [`StateNotifierProvider`](https://pub.dev/documentation/riverpod/latest/riverpod/StateNotifierProvider-class.html) it says:
> 
> > This provider is used in combination with `package:state_notifier`.

You don't have to import it. Riverpod exports it for you. Riverpod's `StateNotifier` _is_ state_notifier's `StateNotifier`

> There aren't any examples for providers with mutable state, where I would manually trigger an update

That's because this is not recommended by the library.

With Riverpod I would like to push immutable state, because this is more maintainable.

It supports ChangeNotifier for backward compatibility (as a mean to help the transition between Provider and Riverpod). But you should avoid it.

> Sometimes you do watch(provider) and sometimes watch(provider.state). Maybe I missed it but I didn't see any explanation of why.

Eh I thought I mentioned it in the doc. I'll add it.

>  Like I said, it seems like a great project and I really want to learn how to use it but there's definitely a slightly inconvenient focus on "how to do a counter" and "how to do a todo list" rather than "this is how it works" and "how to do an app with state that is more complicated than just a list or an integer".


It's difficult. The issue is, what people expect from a doc vary between person. Some like you want a "how it works". Some want a step by step tutorial.

Overall a major issue is, I don't have enough time to properly work on the documentation. All the time that I spend writing doc is time that I could spend developing other things.  
That's especially so because you can find tutorials about Riverpod written by third-party.

In the ideal world, Riverpod would have other maintainers to help me with those. But considering there's no monetary reward, it's difficult to find help.

**Comment by Timmmm** [comment]
Thanks for the reply! That all make sense and I totally appreciate the time constraints.

> With Riverpod I would like to push immutable state, because this is more maintainable.

Ah I wondered if that might be the case. I find the idea of immutable state a bit awkward in languages like Dart & Javascript where everything is a reference, pretty much everything is mutable by default and it's really difficult to control when things are copied (e.g. there's no deep copy function in either).

Fine in C++ or Rust where you can provide a proper optimised data structure that does cheap copy-on-write and so on, so it appears to be immutable but actually does things efficiently behind the scenes. But anyway...

Here's the real app I'm writing (it runs a HTTP server that returns your location). How would you modify it to use Riverpod idiomatically? (I only started it today; it probably doesn't close everything properly).

```
class ServerInfo {
  bool running = false;
  String address = "";
  int port = 0;
  LocationData currentLocation;
}

class ServerNotifier extends StateNotifier<ServerInfo> {
  ServerNotifier() : super(ServerInfo());

  HttpServer _server;
  Location _location;
  StreamSubscription<LocationData> _locationSub;

  startServer() async {
    if (state.running) {
      return;
    }
    // Update state.
    state.running = true;
    state = state;

    // Enable background mode. Initialize requests permissions for wakelocks etc.
    // It does nothing if permission is already granted.
    bool backgroundExecutionInitialized = await FlutterBackground.initialize(androidConfig: backgroundConfig);
    if (!backgroundExecutionInitialized) {
      state.running = false;
      state = state;
      return;
    }
    bool backgroundExecutionEnabled = await FlutterBackground.enableBackgroundExecution();
    if (!backgroundExecutionEnabled) {
      state.running = false;
      state = state;
      return;
    }

    // Start listening to the position.
    _location = new Location();

    bool _serviceEnabled;
    PermissionStatus _permissionGranted;

    _serviceEnabled = await _location.serviceEnabled();
    if (!_serviceEnabled) {
      _serviceEnabled = await _location.requestService();
      if (!_serviceEnabled) {
        await FlutterBackground.disableBackgroundExecution();
        state.running = false;
        state = state;
        return;
      }
    }

    _permissionGranted = await _location.hasPermission();
    if (_permissionGranted == PermissionStatus.denied) {
      _permissionGranted = await _location.requestPermission();
      if (_permissionGranted != PermissionStatus.granted) {
        await FlutterBackground.disableBackgroundExecution();
        state.running = false;
        state = state;
        return;
      }
    }

    await _location.enableBackgroundMode();

    _locationSub = _location.onLocationChanged.listen((event) {
      state.currentLocation = event;
      state = state;
    });

    // Start the server.
    _server = await HttpServer.bind(
      InternetAddress.loopbackIPv4,
      4040,
    );

    state.address = _server.address.host;
    state.port = _server.port;
    state = state;

    handleRequests(_server, () => state.currentLocation);
  }

  stopServer() async {
    if (!state.running) {
      return;
    }
    // Update state.
    state.running = false;
    state.address = "";
    state.port = 0;
    state = state;

    if (_server != null) {
      _server.close();
      _server = null;
    }
    if (_locationSub != null) {
      _locationSub.cancel();
      _locationSub = null;
    }
    if (_location != null) {
      _location = null;
    }
    await FlutterBackground.disableBackgroundExecution();
  }
}

final serverProvider = StateNotifierProvider((ref) => ServerNotifier());
```

I'm guessing something like this:?

```
class ServerInfo {
  ServerInfo(this._running, this._address, this._port, this._currentLocation);
  bool _running = false;
  bool get running => _running;
  String _address = "";
  String get address => _address;
  int _port = 0;
  int get port => _port;
  LocationData _currentLocation;
  LocationData currentLocation => _currentLocation; // Immutability is already broken here because you can modify the contents of LocationData. There's no (easy) way to return a copy.
}
```

And then instead of:

```
    state.running = true;
    state = state;
```

You do

```
    state = ServerInfo(true, state.address, state.port, state.currentLocation);
```

That seems really tedious so I feel like I'm missing something.

**Comment by rrousselGit** [comment]
> And then instead of:
> 
> ```
>     state.running = true;
>     state = state;
> ```
> 
> You do
> 
> ```
>     state = ServerInfo(true, state.address, state.port, state.currentLocation);
> ```
> 
> That seems really tedious so I feel like I'm missing something.

My recommendation is to use StateNotifier in combination with [freezed](https://github.com/rrousselgit/freezed).

It's another package of mine, that allows cloning objects

So you can do things like:

```dart
state = state.copyWith(running: true);
```

**Comment by rrousselGit** [comment]
Another possibility is to have an assert in debug mode that prevents reading `.state` inside `ConsumerWidget.build`

**Comment by TimWhiting** [comment]
@rrousselGit 
I might take a look into improving the documentation. I expect to have some time this coming week I can work on it.

**Comment by felixblaschke** [comment]
I also noticed that the documentation isn't beginner focused, yet.
Therefore I wrote an introduction article: https://felixblaschke.medium.com/riverpod-simplified-an-introduction-to-flutters-most-advanced-state-management-package-c698b4d5a019

Feel free to use it or extract parts to explain Riverpod concept to a beginner audience.

**Comment by Timmmm** [comment]
> My recommendation is to use StateNotifier in combination with [freezed](https://github.com/rrousselgit/freezed).
> 
> It's another package of mine, that allows cloning objects
> 
> So you can do things like:
>
>      state = state.copyWith(running: true);

Ah ok, I think that's the key thing that is missing from the docs! I don't think it's used in any of the examples I could find either.

> Therefore I wrote an introduction article: https://felixblaschke.medium.com/riverpod-simplified-an-introduction-to-flutters-most-advanced-state-management-package-c698b4d5a019

Good work, though I have to say it doesn't cover anything that isn't already covered by the docs, and it didn't address any of the questions I had here. Maybe in the next instalment? :-D

**Comment by felixblaschke** [comment]
> Good work, though I have to say it doesn't cover anything that isn't already covered by the docs, and it didn't address any of the questions I had here. Maybe in the next instalment? :-D

I thought this was a generic issue about the quality of documentation.

Sure everything is somehow covered by the docs but it's spread everywhere. The readme uses StateNotifierProvider with ConsumerWidget, some decision diagram hints to use Consumer. The simple counter uses StateProvider along wit Consumer. The concept page of providers on riverpod.dev uses StreamProvider.

There is no common thread inside the documentation, starting from the most simple usage, guiding towards more advanced techniques (like i.e. composition).

For a new user Riverpod looks very complex with many different things. This is why I recommended beginners to start exactly with two simple things (Consumer + StateProvider). The documentation should also do that, but also hint that there is more to discover.

**Comment by maks** [comment]
I ran into this somewhat just now. I wanted to use Riverpod for a pure Dart project, but the documentation for that use case is quite minimal. I'd be happy to contribute some more documentation for that @rrousselGit  but I'm not even sure I correctly understand what the correct (or your recommended) approach is to then write documentation for. 

For instance how should `ProviderContainer` be used for the pure Dart case? Is there supposed to be a single one (or a tree of them)  that is then passed through to different classes the way build context is in Flutter? 

Sorry if this is too off-topic for this issue, if it is let me know, and I'll raise a separate issue for this.

**Comment by rrousselGit** [comment]
@maks I'd suggest raising a new issue indeed

Dart-only projects need more than just documentation. They most likely need some extra APIs to make things easier, but that needs to be discussed as I don't fully understand the use-case.

**Comment by kevlar700** [comment]
> > There aren't any examples for providers with mutable state, where I would manually trigger an update
> 
> That's because this is not recommended by the library.
> 
> With Riverpod I would like to push immutable state, because this is more maintainable.
> 
> It supports ChangeNotifier for backward compatibility (as a mean to help the transition between Provider and Riverpod). But you should avoid it.

I shall review statenotifier for a third time. However my conclusion, so far has been that I like explicit builds and I believe changenotifier results in more readable and flexible code and rate those as higher importance than immutability in the state class.

To be brutally honest. I actually find the examples on the homepage to be quite unreadable.

**Comment by YoofiBP** [comment]
I'm here a few months later to reiterate the points @Timmmm mentioned in his original issue post as I tried to use Riverpod for the first time today. 

The docs are quite hard for beginners with Riverpod to follow as they mention different Providers and methods (Providers, FutureProviders, autoDispose, context.read) without a guide as to how it all works for the use cases.  

I am aware @rrousselGit  is currently pressed for time to make changes to the docs. Does anyone have a link to any beginner-friendly resources to help understand Riverpod. I checked out @felixblaschke article and found it helpful and would like to know if other resources also exist beyond combing through the current docs.

**Comment by ojm-it** [comment]
Is there any progress being made on this issue?

**Comment by bizz84** [comment]
Not sure if this is of any help, but I have an updated guide to Riverpod here:

https://codewithandrea.com/videos/flutter-state-management-riverpod/

It doesn’t cover as much stuff as the documentation, but it follows a linear structure.

**Comment by maks** [comment]
@rrousselGit I think you do a great job with the Riverpod documentation, but @bizz84 I consider your article to be the _unofficial_ Riverpod manual! Personally I found it _**invaluable**_ for getting to grips with Riverpod. I find reading documentation with a specific application example like yours very helpful and would humbly suggest that the article be linked in the Riverpod Readme.

The only gap I've found in both the Riverpod documentation and @bizz84's tutorial is an example of how to get a reference to a `StateNotifierProvider` rather the `state` object its notifying about. What I mean by that is that I tend to use `StateNotifierProvider`'s as "viewmodels" so just as in [@bizz84 tutorial on them](https://codewithandrea.com/videos/flutter-state-management-setstate-freezed-state-notifier-provider/) I would have one that has methods on it like in the article: 
eg.
```
class CreateProfileModel extends StateNotifier<CreateProfileState> {
  CreateProfileModel({@required this.dataStore})
      : super(const CreateProfileState.noError());
  final DataStore dataStore;

  Future<bool> createProfile(String name) async {
   ...
  }
}
```

and then somewhere else is say a `onPress()` I would need to get a ref to it to call the `createProfile()`. In Provider, this per the article, this would be:
```
final model = context.read<CreateProfileModel>();
model.creatProfile('...');
```

whereas in Riverpod I figured out I need to instead do:
```
 final model = ref.read(createProfileModelProvider.notifier);
```

Not sure if it's me being slow, but that `.notifier` part was something I couldn't easily find in the Riverpod tutorial docs and took me a little while to figure out on my own.

**Comment by rrousselGit** [comment]
Documentation rework hasn't quite started yet on my side as I've been busy with the v1 release – which is coming today.

With that out of the way, the next steps are documentation and the Riverpod devtool, so expect some changes soon.

**Comment by ntc2** [comment]
I completely agree with / had some of these same questions as the OP:

> 5. There aren't any examples for providers with mutable state, where I would manually trigger an update.

I see `ChangeNotifierProvider` mentioned in the `Concepts -> Providers`, but not in the `API Reference -> All Providers` in the nav bar on the side of the docs. If this is intentional, I think it would be more helpful to also mention `ChangeNotifierProvider` in the `All Providers`. I understand that you prefer using immutable state, and I'm all on board with that after 5+ years of professional Haskell programming in the past, but like the OP says in another comment above, I find that mutable state + manual notification is sometimes a better fit for Dart, for me. I wish Dart had better support for immutability (namely data classes / records without boilerplate, and immutable containers with efficient modification), but in the mean time, it seems simpler (less code, more efficient, and not any more error prone) to e.g. just hide a mutable array inside a `ChangeNotifier` sometimes.

> 6. Sometimes you do `watch(provider)` and sometimes `watch(provider.state)`. Maybe I missed it but I didn't see any explanation of why.

In my case, I got really confused by the ["don't use `ref.read` in `build`" example in the docs](https://riverpod.dev/docs/concepts/reading#dont-use-refread-inside-the-build-method). After looking at it enough times, I realized that the point was probably that you're `read`/`watch`ing `provider.notifier` and not plain `provider`, but an explanation of when or why you'd want to do that would be helpful. Reading the docs up to that point, my impression was that you always do `read(provider)`, and so at first I didn't notice the difference, and then had a hard time guessing why it worked / what it meant.

7. I'd appreciate some more explanation of how it all works up front. I suspect it's pretty much impossible to use this without understanding at least vaguely how it works, and the current guide is a load of "if you want this, then do this" rather than "when you call watch() it does this".

Specifically, a prominent explanation of when `ref.watch(thing)` will trigger a rebuild would be really helpful. For example, from the docs it was not clear to me when a provider like `boolProvider` here would notify its listeners:

```
final boolProvider = Provider.family<bool, String>((ref, name) {
  final set = ref.watch(setProvider);
  return set.contains(name);
});
```

Some experimentation showed me that this works the way I would hope, i.e. changes to `setProvider` trigger a rebuild of all the providers in the `boolProvider` family, but those providers don't themselves notify their listeners *unless* their computed `bool` actually changes. But, before doing the experiment, it seemed entirely possible to me that all the `ref.watch(boolProvider(name))` listeners would be notified, since the docs didn't say anything about it either way. It seems what's going on here is that when `boolProvider(name)` gets rebuilt, Riverpod compares the old and new output using `==`, and only notifies listeners on a change, but it would be helpful if the docs actually said that.
